src/net.cpp

```diff
@@ -802,7 +802,13 @@ size_t CConnman::SocketSendData(CNode& node) const
             if (!node.m_sock) {
                 break;
             }
-            nBytes = node.m_sock->Send(reinterpret_cast<const char*>(data.data()) + node.nSendOffset, data.size() - node.nSendOffset, MSG_NOSIGNAL | MSG_DONTWAIT);
+            int flags = MSG_NOSIGNAL | MSG_DONTWAIT;
+#ifdef MSG_MORE
+            if (it + 1 != node.vSendMsg.end()) {
+                flags |= MSG_MORE;
+            }
+#endif
+            nBytes = node.m_sock->Send(reinterpret_cast<const char*>(data.data()) + node.nSendOffset, data.size() - node.nSendOffset, flags);
```

Comment:
If this can't send the full message one of the two `break`s below will be executed which may result in this function returning with all `Send()` calls using `MSG_MORE`. For example: let there be 3 messages, 10 bytes each. `Send()` of the first one will use `MSG_MORE`, let it return 10 (full message sent). `Send()` of the second one will use `MSG_MORE`, let it return any number in [0..9]. Then this function will return, without ever calling `Send()` without `MSG_MORE` and the first message may remain buffered/unsent somewhere in the OS.

I think this is unwanted effect, defeating `TCP_NODELAY` and the check `if (it + 1 != node.vSendMsg.end())`.

I wonder if it is possible to tell the OS that we are done, without sending further data? Maybe `Send(nullptr, 0, MSG_NOSIGNAL | MSG_DONTWAIT); // without MSG_MORE` would do that?

Reply:
If not all of our pending outgoing messages can fit in the socket's transmit buffer, then we *want* all of our calls to `send` to specify `MSG_MORE`. The kernel will immediately (without delay) transmit as many full segments as it can assemble and that will fit in the congestion window. It will not immediately transmit a partial segment but instead will call back to us (via returning a writable status from `select`, `poll`, `epoll`, etc.) to fetch more data to transmit. If we have no more data to transmit, then yes, calling `send` with a zero length and without `MSG_MORE` will tell the kernel that we have nothing left to send, and it will transmit a partial segment as soon as such segment will fit in the congestion window.

In your example, if the second of the three 10-byte messages can only partially be copied to the socket transmit buffer, then it must be the case that the transmit buffer is now full. The transmit buffer is always larger than MSS, the maximum segment size, meaning the kernel can now assemble a full segment, which it will transmit as soon as the window allows. It will then indicate the socket is writable again, and we will be called again to write more of the second message.

So, in short, your concern that we will unnecessarily delay transmitting data is unwarranted. The kernel never delays transmitting a full segment if it fits in the congestion window. The kernel does delay transmitting a partial segment that fits in the congestion window if the last call to `send` on that socket specified `MSG_MORE`, but that's exactly what we want, as we want the opportunity to write more data into the transmit buffer so the kernel will potentially transmit fewer and larger segments.

Reply:
If `send(2)` writes less bytes than requested, does it follow that the transmit buffer is full? I am not sure about that. Is it documented somewhere? What about a signal interrupting the send midway (or even before it started)? Is that not a case where `send(2)` writes less bytes than requested, but the transmit buffer is not full?

Reply:
> What about a signal interrupting the send midway (or even before it started)?

I'm pretty sure that's only possible if the socket is in blocking mode. But regardless, if `send` returns a short write due to invoking a signal handler, then the socket is still writable, and the user code will call `send` again without delay (i.e., on the next iteration of the I/O loop).

Reply:
That would be after accepting new connections, sending, receiving and parsing messages from all other (possibly 100s) connections. I guess this would be rare and even when it happens would probably have no adverse effects on the communication, but I would feel more comfortable if this PR does not introduce that. If a dummy `Send()` without `MSG_MORE` before the two `break`s is a way to avoid that, then I think it should be added.

Reply:
In no case will the kernel hold onto a partial segment in the socket's transmit buffer for longer than 250 ms if the segment fits in the congestion window, so that's the maximum latency that could be introduced by writing a complete message with `MSG_MORE`, maxing out the transmit buffer while writing a subsequent message also with `MSG_MORE`, and then going off to do other things for a nigh eternity before coming back around.

Anyway, I don't want to add *more* syscall overhead. I'd prefer to rewrite this to do gathering writes rather than adding yet another `send` call. If that's the only way you'll let this move forward, then I'll do that work, but then I'd expect that to be bike shedded for ages while people debate the merits of stack allocation versus heap allocation of the `struct iovec` array, platform agnosticism, abstraction leakage, readability and maintainability of code, and scope creep.

Reply:
> > ... 250 ms ...
> 
> That seems like relying on an undocumented behavior (unless it is documented somewhere)?

It is documented. (It's actually 200 milliseconds. I misremembered.)

[`sendmsg(2)`](https://linux.die.net/man/2/sendmsg) says:

> **MSG_MORE** (Since Linux 2.4.4)
> The caller has more data to send. This flag is used with TCP sockets to obtain the same effect as the **TCP_CORK** socket option (see `tcp(7)`), with the difference that this flag can be set on a per-call basis.

[`tcp(7)`](https://linux.die.net/man/7/tcp) says:

> **TCP_CORK** (since Linux 2.2)
> If set, don't send out partial frames. All queued partial frames are sent when the option is cleared again. This is useful for prepending headers before calling `sendfile(2)`, or for throughput optimization. As currently implemented, there is a 200 millisecond ceiling on the time for which output is corked by **TCP_CORK**. If this ceiling is reached, then queued data is automatically transmitted.

Reply:
The kernel will never wait to send full segments.  Vasil's concern is the case where we try to send a non-final message, run out of space in the kernel's transmit buffer (so only some of the message was copied in), then go off and do other things for a virtual eternity. If we haven't called `send` without `MSG_MORE`, then it is *possible* that there is a complete message sitting in the kernel's transmit buffer (ahead of the incomplete message that we failed to fully write) that will be delayed. That's true, but I'd say that's by design. If it's really taking us longer than 200 ms for the network I/O loop to make a full pass, we have much bigger problems!

Reply:
The (very reasonable) hope is that the network I/O loop comes back around to our socket to finish writing the incomplete message (and possibly more messages) before the kernel's transmit buffer has drained below the size of a full segment. In that case, there will be no artificial delay introduced to any message we've written to the socket.

So, basically, we want the network I/O loop to come back around to service our socket again in less time than the kernel's transmit buffer takes to drain.

Reply:
> It is documented

Even better, thanks!

Reply:
> 221 times a `MSG_MORE` was interrupted

I think what you may have meant was that a `send` call indicated that fewer bytes were copied than were requested, not that the call was interrupted (by a signal handler).

The real test would be querying the `SIOCOUTQ` ioctl on the socket the *next* time the network I/O loop gets around to servicing it. That's the number of untransmitted bytes sitting in the kernel's transmit buffer. If that number is smaller than the connection's MSS, then the kernel exhausted its transmit buffer and started to delay sending the remaining bytes due to `MSG_MORE`.

Reply:
> never with having sent nothing

*N.B.:* It would be impossible for zero bytes to have been copied to a socket in non-blocking mode if that socket had indicated that it was writable and no other thread had written to it since such indication was made. Again, I'm pretty sure that even signal handlers won't interrupt a non-blocking write since the kernel only enters an interruptible wait state during a blocking write.